### PR TITLE
Fix for list in mixins applied twice

### LINKED ISSUE #1649: SNOW-1705656: Lists in mixins get applied twice
### SnowCLI version

main

### Python version

Python 3.12.3

### Platform

macOS-14.6.1-arm64-arm-64bit

### What happened

Using mixins that contain lists of dicts, like `artifacts`, the list is applied twice. In this minimal example
```yml
definition_version: 2
entities:
  pkg:
    type: application package
    manifest: app/manifest.yml
    meta:
      use_mixins:
        - artifacts_mixin
mixins:
  artifacts_mixin:
    artifacts:
      - src: app/*
        dest: ./
```
The resultant project model is (irrelevant fields omitted)
```yml
definition_version: '2'
entities:
  pkg:
    type: application package
    artifacts:
    - src: app/*
      dest: ./
      processors: []
    - src: app/*
      dest: ./
      processors: []
    manifest: app/manifest.yml
```

I believe this is because `_unique_extend` doesn't properly exclude list items that are dicts when defaults have been added, since the items are no longer identical:
image

### Console output

_No response_

### How to reproduce

Evaluate the above snowflake.yml file in a debugger to see that the list is duplicated

### PR DESCRIPTION
**Pull request by sfc-gh-jsikorski**
### Pre-review checklist
   * [ ] I've confirmed that instructions included in README.md are still correct after my changes in the codebase.
   * [ ] I've added or updated automated unit tests to verify correctness of my new code.
   * [ ] I've added or updated integration tests to verify correctness of my new code.
   * [ ] I've confirmed that my changes are working by executing CLI's commands manually on MacOS.
   * [ ] I've confirmed that my changes are working by executing CLI's commands manually on Windows.
   * [ ] I've confirmed that my changes are up-to-date with the target branch.
   * [ ] I've described my changes in the release notes.
   * [ ] I've described my changes in the section below.

### Changes description
Fixes #1649

### CODE REVIEW

File: src/snowflake/cli/api/project/schemas/project_definition.py [hunk 1]
```diff
@@ -209,11 +212,13 @@ def _merge_mixins_with_entity(
         for mixin_name in entity_mixins_names:
             if mixin_name not in mixin_defs:
                 raise ValueError(f"Mixin {mixin_name} not defined")
-
+        context = get_cli_context()
```
Comment (sfc-gh-turbaszek) [inline src/snowflake/cli/api/project/schemas/project_definition.py:215]: Why we need context here?

File: src/snowflake/cli/api/project/schemas/project_definition.py [hunk 2]
```diff
@@ -209,11 +212,13 @@ def _merge_mixins_with_entity(
...
         # Build object override data from mixins
         data: dict = {}
         for mx_name in entity_mixins_names:
-            data = cls._merge_data(data, mixin_defs[mx_name])
+            if mx_name not in cls._applied_mixins.default.get(entity_id, []):  # type: ignore
```
Comment (sfc-gh-turbaszek) [inline src/snowflake/cli/api/project/schemas/project_definition.py:219]: What is `cls._applied_mixins.default`? Isn't `cls._applied_mixins` a dictionary?

File: src/snowflake/cli/api/project/schemas/project_definition.py (hunk 1 above)
Comment (sfc-gh-jsikorski) [reply src/snowflake/cli/api/project/schemas/project_definition.py:215]: Leftover from previous attempts, already removed

File: src/snowflake/cli/api/project/schemas/project_definition.py (hunk 2 above)
Comment (sfc-gh-jsikorski) [reply src/snowflake/cli/api/project/schemas/project_definition.py:219]: Pydantic wraps it in its' own `ModelPrivateField` class, that needs using this access methods

Comment (sfc-gh-jsikorski) [reply src/snowflake/cli/api/project/schemas/project_definition.py:220]: Added docstring in attribute declaration

**Pull request closed without merging**